PhysicsTools/NanoAOD/plugins/LHETablesProducer.cc

```diff
@@ -0,0 +1,98 @@
+            auto lheTab  = std::make_unique<nanoaod::FlatTable>(1, "LHE", true);
+
+            edm::Handle<LHEEventProduct> lheInfo;
+            if (iEvent.getByToken(lheTag_, lheInfo)) {
+                fillLHEObjectTable(*lheInfo, *lheTab);
+            }
+
+            iEvent.put(std::move(lheTab));
+        }
+
+        void fillLHEObjectTable(const LHEEventProduct & lheProd, nanoaod::FlatTable & out) const {
+            double lheHT = 0, lheHTIncoming = 0;
+            unsigned int lheNj = 0, lheNb = 0, lheNc = 0, lheNuds = 0, lheNglu = 0;
+            double lheVpt = 0;
+
+            const auto & hepeup = lheProd.hepeup();
+            const auto & pup = hepeup.PUP;
+            int lep = -1, lepBar = -1, nu = -1, nuBar = -1;
+            for (unsigned int i = 0, n = pup.size(); i  < n; ++i) {
+                int status = hepeup.ISTUP[i];
+                int idabs = std::abs(hepeup.IDUP[i]);
+                if ( (status == 1) && ( ( idabs == 21 ) || (idabs > 0 && idabs < 7) ) ) { //# gluons and quarks
+                    // object counters
+                    lheNj++;
+                    if (idabs==5) lheNb++;
+                    else if (idabs==4) lheNc++;
+                    else if (idabs <= 3 ) lheNuds++;
+                    else if (idabs == 21) lheNglu++;
+                    // HT
+                    double pt = std::hypot( pup[i][0], pup[i][1] ); // first entry is px, second py
+                    lheHT += pt;
+                    int mothIdx = std::max(hepeup.MOTHUP[i].first-1,0); //first and last mother as pair; first entry has index 1 in LHE; incoming particles return motherindex 0
+                    int mothIdxTwo = std::max(hepeup.MOTHUP[i].second-1,0);
+                    int mothStatus  = hepeup.ISTUP[mothIdx];
+                    int mothStatusTwo  = hepeup.ISTUP[mothIdxTwo];
+                    bool hasIncomingAsMother = mothStatus<0 || mothStatusTwo<0;
+                    if (hasIncomingAsMother) lheHTIncoming += pt;
+                } else if (idabs == 12 || idabs == 14 || idabs == 16) {
+                    (hepeup.IDUP[i] > 0 ? nu : nuBar) = i;
+                } else if (idabs == 11 || idabs == 13 || idabs == 15) {
+                    (hepeup.IDUP[i] > 0 ? lep : lepBar) = i;
+                }
+            }
+            std::pair<int,int> v(0,0);
+            if      (lep != -1 && lepBar != -1) v = std::make_pair(lep,lepBar);
+            else if (lep != -1 &&  nuBar != -1) v = std::make_pair(lep, nuBar);
+            else if (nu  != -1 && lepBar != -1) v = std::make_pair(nu ,lepBar);
+            else if (nu  != -1 &&  nuBar != -1) v = std::make_pair(nu , nuBar);
+            if (v.first != -1 && v.second != -1) {
+                lheVpt = std::hypot( pup[v.first][0] + pup[v.second][0], pup[v.first][1] + pup[v.second][1] ); 
+            }
+
+            out.addColumnValue<uint8_t>("Njets", lheNj, "Number of jets (partons) at LHE step", nanoaod::FlatTable::UInt8Column);
```

Comment:
it doesn't look normal that every event we have to specify types and documentation description string.

Reply:
it's an overhead but it gives us the flexibility we need (eg allowing the schema to depend on the configuration) while keeping the output module generic and the code simple (avoid percolating stuff through provenance or run/lumi branches).
we may re-engineer things in the future if the overhead becomes large  (now i/o and other physics object code dominate the costs)

Reply:
It's not just the overhead: this implementation apparently allows changes in format event by event, which sounds dangerous.

Reply:
the nanoaod format is by design less structured.
We made some design choice balancing risks and benefits. We would be happy to discuss this design in any of the XPOG meeting where the NanoAOD format is discussed.

Reply:
I'm not sure what this response means.
Does the Nano format support event-by-event changes in branch/table content?
Please clarify.

Reply:
in some cases (not in the one above) it does. E.g. HLT bits are allowed to change from one event to another and proper (back)filling of the branches is implemented.
As for the rest, while it is not meant to change event by event, there is no "configuration" of the event content specified explicitelly anywhere (not at compile time in the DataFormats, not explicitely at config time). 
The configuration of the branches to write is configured basically at first event. 

I do not think this discussion belong to a github PR. If you want to understand why we designed it so, we can have a chat on this topic.